README.md
<p align="center">
  <img src="https://github.com/wipodev/ColabKit/blob/main/ColabKit.png" alt="ColabKit logo" width="250" height="250">
</p>

# ColabKit

ColabKit is a Python library designed to enhance the experience of working in Google Colab environments. With ColabKit, you can simplify common tasks, manipulate media, record audio, and create interactive widgets more efficiently.

## Installation

You can easily install ColabKit using `pip`. Open a terminal and run:

```bash
pip install ColabKit
```

## Test in google colab

You can see the tests carried out on the library in Google Colab through this link

<a href="https://colab.research.google.com/github/wipodev/ColabKit/blob/main/Tests_ColabKit.ipynb" target="_blank"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Key Features

- **Media Manipulation**: Visualize and edit video and audio files directly in your notebook.
- **Audio Recording**: Record real-time audio through your browser with a simple function.
- **Interactive Widgets**: Create interactive buttons in the notebook to execute custom code.
- **System Utilities**: Simplify common tasks, such as file deletion and file existence checks.

## Usage Examples

### Video Display

```python
from ColabKit.media import showMedia

# Display a video file
showMedia('video.mp4', 'video')
```

### Audio Recording

```python
from ColabKit.record import record

# Record audio and save the result to a WAV file
record('recording.wav')
```

### Creating Interactive Buttons

```python
from ColabKit.widget import button

# Create an interactive button
def my_function():
    print("You clicked the button!")

button("Click Me", "Click to execute my_function", "success", my_function)
```

## Documentation

For more information and detailed documentation, visit [the ColabKit repository on GitHub](https://github.com/wipodev/ColabKit).

## Contributions

We appreciate contributions! If you wish to contribute or report issues, visit [our GitHub repository](https://github.com/wipodev/ColabKit).

## License

This project is under the MIT License. See the [LICENSE](https://github.com/wipodev/ColabKit/blob/main/LICENSE) file for more details.

---


CHANGELOG.md
# Inaugural Release

This is the first official release of ColabKit, a Python library designed to streamline work in Google Colab environments. ColabKit offers a collection of modules, including tools for handling media, recording audio, managing files, and creating interactive widgets in your Colab notebooks.

**Key Highlights:**

- `media`: Visualize and process media in your Colab notebooks.
- `record`: Record and manage audio recordings directly from your browser.
- `utils`: Utilities for file management and interaction with Google Colab.
- `widget`: Create interactive buttons and widgets in your notebooks.

We are thrilled to share this initial version with the Python and Colab community. We hope ColabKit aids your projects and experiments in Google Colab. Please feel free to provide feedback, report issues, or contribute to the library's development on our GitHub repository.

**Documentation and Usage:**

For more details on how to use ColabKit and access the complete documentation, please visit [link to documentation](https://github.com/wipodev/ColabKit).

**Additional Notes:**

- Be sure to keep your ColabKit installation up to date using `pip install ColabKit --upgrade`.
- If you have any questions or issues, don't hesitate to open an issue on our GitHub repository for support.

Thank you for your support and enthusiasm in the development of ColabKit. We hope you find this library useful in your projects!


tests/test_media.py
import unittest
from unittest.mock import patch, MagicMock
from src.ColabKit import media
import ffmpeg


class MediaTestCase(unittest.TestCase):

    @patch('src.ColabKit.media.ffmpeg.probe')
    def test_get_duration_valid(self, mock_probe):
        # Configura el retorno simulado de la función ffmpeg.probe
        mock_probe.return_value = {'format': {'duration': '10.5'}}
        video_file = 'video.mp4'
        duration = media.get_duration(video_file)
        self.assertIsInstance(duration, float)
        self.assertEqual(duration, 10.5)

    @patch('src.ColabKit.media.ffmpeg.probe')
    def test_get_video_resolution_valid(self, mock_probe):
        # Configura el retorno simulado de la función ffmpeg.probe
        mock_probe.return_value = {'streams': [
            {'codec_type': 'video', 'width': 1920, 'height': 1080}]}
        video_file = 'video.mp4'
        resolution = media.get_video_resolution(video_file)
        self.assertIsInstance(resolution, tuple)
        self.assertEqual(resolution, (1920, 1080))

    def test_formatted_err(self):
        # Test case for an error message from FFmpeg
        stderr_output = b"Error: Something went wrong\nAnother comment occurred"
        formatted_output = media.get_formatted_error(stderr_output)
        expected_output = "\033[43m\033[101mError: Something went wrong\033[0m"
        # Assert that the error message is correct after formatting
        self.assertEqual(formatted_output, expected_output)

    @patch('ffmpeg.run')
    @patch('builtins.print')
    @patch('src.ColabKit.media.remove_file')
    def test_run_proc_success(self, mock_remove_file, mock_print, mock_run):
        # Test case for a successful FFmpeg process
        mock_run.return_value = None
        result = media.run_process('output_stream', 'output_file', 'message')
        # Assert that the function returned True after executing the FFmpeg process
        self.assertTrue(result)
        # Assert that the print function was called with the expected message
        mock_print.assert_called_with('File message and saved as output_file')
        # Assert that the remove_file function was not called
        mock_remove_file.assert_not_called()

    @patch('ffmpeg.run', side_effect=ffmpeg.Error("", "", stderr=b'Error occurred'))
    @patch('builtins.print')
    @patch('src.ColabKit.media.remove_file')
    def test_run_proc_failure(self, mock_remove_file, mock_print, mock_run):
        # Test case for a failed FFmpeg process
        result = media.run_process('output_stream', 'output_file', 'message')
        # Assert that the function returned False after executing the FFmpeg process
        self.assertFalse(result)
        # Assert that the print function was called with the expected message
        mock_print.assert_called_with(
            media.get_formatted_error(b'Error occurred'))
        # Assert that the remove_file function was called
        mock_remove_file.assert_called_with('output_file')

    @patch('src.ColabKit.media.run_process', return_value=True)
    @patch('builtins.print')
    def test_conv_success(self, mock_print, mock_run_process):
        # Test case for a successful conversion
        result = media.convert('input_file.mp4', 'output_file.mp4')
        # Assert that the function returned True after executing the FFmpeg process
        mock_print.assert_called_with(
            "Converting input_file.mp4 to output_file.mp4...")
        # Assert that the remove_file function was called
        self.assertTrue(result)

    @patch('src.ColabKit.media.run_process', return_value=False)
    @patch('builtins.print')
    def test_conv_failure(self, mock_print, mock_run_process):
        # Test case for a failed conversion
        result = media.convert('input_file.mp4', 'output_file.mp4')
        # Assert that the function returned False after executing the FFmpeg process
        mock_print.assert_called_with(
            "Converting input_file.mp4 to output_file.mp4...")
        # Assert that the remove_file function was called
        self.assertFalse(result)

    @patch('src.ColabKit.media.run_process', return_value=True)
    @patch('builtins.print')
    def test_short_success(self, mock_print, mock_run_process):
        # Test case for a successful trim
        result = media.shorten('input_file.mp4', 10, 5, 'output_file.mp4')
        # Assert that the function returned True after executing the FFmpeg process
        mock_print.assert_called_with(
            "shorting input_file.mp4 to 5 seconds...")
        # Assert that the remove_file function was called
        self.assertTrue(result)

    @patch('src.ColabKit.media.run_process', return_value=False)
    @patch('builtins.print')
    def test_short_failure(self, mock_print, mock_run_process):
        # Test case for a failed trim
        result = media.shorten('input_file.mp4', 10, 5, 'output_file.mp4')
        # Assert that the function returned False after executing the FFmpeg process
        mock_print.assert_called_with(
            "shorting input_file.mp4 to 5 seconds...")
        # Assert that the remove_file function was called
        self.assertFalse(result)

    @patch('src.ColabKit.media.get_video_resolution', return_value=(1920, 1080))
    @patch('src.ColabKit.media.run_process', return_value=True)
    @patch('builtins.print')
    def test_resize_video_success(self, mock_print, mock_run_process, mock_get_video_resolution):
        # Test case for a successful resize
        result = media.resize_video('input_video.mp4', 720)
        # Assert that the function returned True after executing the FFmpeg process
        mock_print.assert_called_with("Resizing video to 720p...")
        # Assert that the remove_file function was called
        self.assertEqual(result, 'input_video_720p.mp4')

    @patch('src.ColabKit.media.get_video_resolution', return_value=(1280, 720))
    @patch('src.ColabKit.media.run_process', return_value=True)
    @patch('builtins.print')
    def test_resize_video_small_resolution(self, mock_print, mock_run_process, mock_get_video_resolution):
        # Test case for a successful resize
        result = media.resize_video('input_video.mp4', 720)
        # Assert that the function returned True after executing the FFmpeg process
        mock_print.assert_called_with("No need to change the size")
        # Assert that the remove_file function was called
        self.assertEqual(result, 'input_video.mp4')

    @patch('src.ColabKit.media.get_video_resolution', return_value=(640, 360))
    @patch('builtins.print')
    def test_resize_video_no_need_to_change(self, mock_print, mock_get_video_resolution):
        # Test case for a successful resize
        result = media.resize_video('input_video.mp4', 720)
        # Assert that the function returned True after executing the FFmpeg process
        mock_print.assert_called_with("No need to change the size")
        # Assert that the remove_file function was called
        self.assertEqual(result, 'input_video.mp4')
        # Assert that the get_video_resolution function was called
        mock_get_video_resolution.assert_called_once_with('input_video.mp4')

    @patch('src.ColabKit.media.get_video_resolution', return_value=(1920, 1080))
    @patch('src.ColabKit.media.run_process', return_value=False)
    @patch('builtins.print')
    def test_resize_video_failure(self, mock_print, mock_run_process, mock_get_video_resolution):
        # Test case for a failed resize
        result = media.resize_video('input_video.mp4', 720)
        # Assert that the function returned False after executing the FFmpeg process
        mock_print.assert_called_with("Resizing video to 720p...")
        # Assert that the remove_file function was called
        self.assertFalse(result)


if __name__ == '__main__':
    unittest.main()


tests/test_utils.py
from unittest.mock import patch, MagicMock
from src.ColabKit import utils
import unittest
import os


class UtilsTestCase(unittest.TestCase):

    file_path = "tests/file.txt"
    file_not_exists = "non_existing/file.txt"
    # Create a new file
    with open(f"./{file_path}", 'w') as file:
        pass

    def test_existing_file(self):
        # Test case for an existing file
        result = utils.file_exists(self.file_path)
        self.assertTrue(result)

    def test_non_existing_file(self):
        # Test case for a non-existing file
        result = utils.file_exists(self.file_not_exists)
        self.assertFalse(result)

    def test_directory(self):
        # Test case for a directory
        route = "/tests"
        result = utils.file_exists(route)
        self.assertFalse(result)

    def test_remove_existing_file(self):
        utils.remove_file(self.file_path)
        self.assertFalse(os.path.exists(self.file_path))

    def test_remove_non_existing_file(self):
        # Test removing a non-existing file
        utils.remove_file(self.file_not_exists)
        self.assertFalse(os.path.exists(self.file_not_exists))

    @patch('os.path.exists', side_effect=[False, True, True, True])
    @patch('os.path.getsize', side_effect=[0, 100, 100, 100])
    @patch('time.sleep', new_callable=MagicMock)
    def test_wait_for_existing_unchanged_file(self, mock_sleep, mock_getsize, mock_exists):
        # Test case for an existing file
        result = utils.wait_for_file('test_file.txt', time_limit=3)
        # Assert that the result is True
        self.assertTrue(result)

    @patch('os.path.exists', side_effect=[False, True, True, True])
    @patch('os.path.getsize', side_effect=[0, 100, 200, 200])
    @patch('time.sleep', new_callable=MagicMock)
    def test_wait_for_existing_changed_file(self, mock_sleep, mock_getsize, mock_exists):
        # Test case for an existing file
        result = utils.wait_for_file('test_file.txt', time_limit=3)
        # Assert that the result is True
        self.assertTrue(result)

    @patch('os.path.exists', return_value=False)
    @patch('time.sleep', new_callable=MagicMock)
    def test_wait_for_non_existing_file(self, mock_sleep, mock_exists):
        # Test case for a non-existing file
        result = utils.wait_for_file('non_existing_file.txt', time_limit=3)
        # Assert that the result is False
        self.assertFalse(result)

    @patch('os.path.exists', side_effect=[False, True, True, True])
    @patch('os.path.getsize', side_effect=[0, 100, 100, 100])
    @patch('time.sleep', new_callable=MagicMock)
    def test_wait_for_existing_unchanged_file_with_custom_time_limit(self, mock_sleep, mock_getsize, mock_exists):
        # Test case for an existing file
        result = utils.wait_for_file('test_file.txt', time_limit=2)
        # Assert that the result is True
        self.assertTrue(result)


if __name__ == '__main__':
    unittest.main()


tests/test_widget.py
from unittest.mock import MagicMock, patch
from src.ColabKit.widget import button
import ipywidgets as widgets
import unittest


class WidgetTestCase(unittest.TestCase):

    text = "My Button"
    tip = "Click here"
    style = "success"
    func = MagicMock()
    param = (1, 2, 3)


def test_generate_button_creation(self):
    with patch('ipywidgets.Button', new_callable=MagicMock) as mock_button:
        # Create the button and call the function with the given arguments
        button(self.text, self.tip, self.style,
               self.func, *self.param)
        # Assert that the Button class was called with the expected arguments
        mock_button.assert_called_with(
            description=self.text, button_style=self.style, tooltip=self.tip, icon="erase")

    def test_generate_button_function_execution(self):
        # Create button with specified parameters
        button = button(
            self.text, self.tip, self.style, self.func, *self.param)
        # Simulate button click
        button.click()
        # Verify that the function was called with the expected parameters
        self.func.assert_called_with(*self.param)


if __name__ == '__main__':
    unittest.main()


src/ColabKit/utils.py
from . import GCOLAB
import time
import os

if GCOLAB:
    from google.colab import files, drive


def remove_file(route) -> None:
    if os.path.isfile(route):
        os.remove(route)


def file_exists(route) -> bool:
    return os.path.exists(route)


def mount_drive() -> None:
    if not os.path.isdir("/content/drive/MyDrive"):
        drive.mount('/content/drive', force_remount=True)


def upload_file(route) -> None:
    remove_file(route)
    uploaded = files.upload()
    for filename in uploaded.keys():
        os.rename(filename, route)


def wait_for_file(file_name: str, time_limit: int = 60) -> bool:
    initial_time = time.time()
    exists = False

    while time.time() - initial_time < time_limit:
        if os.path.exists(file_name):
            previous_size = os.path.getsize(file_name)
            time.sleep(1)
            current_size = os.path.getsize(file_name)
            if current_size == previous_size:
                exists = True
                break
    return exists


src/ColabKit/record.py
from .template import AUDIO_RECORD, SAVED_AUDIO_RECORDING
from scipy.io.wavfile import read as wav_read, write as wav_write
from IPython.display import HTML, display
from .utils import wait_for_file
from base64 import b64decode
from . import GCOLAB
import ffmpeg
import io

if GCOLAB:
    from google.colab.output import eval_js


def get_audio() -> tuple:
    display(HTML(AUDIO_RECORD))
    data = eval_js("data")
    binary = b64decode(data.split(',')[1])
    process = (
        ffmpeg
        .input('pipe:0')
        .output('pipe:1', format='wav')
        .run_async(pipe_stdin=True, pipe_stdout=True, pipe_stderr=True, quiet=True, overwrite_output=True)
    )
    output, err = process.communicate(input=binary)
    riff_chunk_size = len(output) - 8
    q = riff_chunk_size
    b = []
    for i in range(4):
        q, r = divmod(q, 256)
        b.append(r)
    riff = output[:4] + bytes(b) + output[8:]
    sr, audio = wav_read(io.BytesIO(riff))
    return audio, sr


def record(file_audio="record.wav") -> None:
    audio, sr = get_audio()
    wav_write(file_audio, sr, audio)
    if wait_for_file(file_audio):
        display(HTML(SAVED_AUDIO_RECORDING))


src/ColabKit/template.py
AUDIO_SHOW = """<audio controls src="%s"></audio>"""

VIDEO_SHOW = """<video controls width=500><source src="%s" type="video/mp4"></video>"""

AUDIO_RECORD = """
<script>
    var my_div = document.createElement("DIV");
    var recordButton = document.createElement("BUTTON");

    my_div.style.display = "flex";
    my_div.style.flexDirection = "column";
    my_div.style.gap = "10px";
    my_div.style.alignItems = "center";
    my_div.style.padding = "10px";

    recordButton.style.backgroundColor = "red";
    recordButton.style.color = "white";
    recordButton.style.margin = "0 auto";
    recordButton.style.border = "none";
    recordButton.style.borderRadius = "5px";
    recordButton.style.padding = "5px 10px";
    recordButton.style.cursor = "pointer";

    my_div.appendChild(recordButton);
    document.body.appendChild(my_div);

    var base64data = 0;
    var reader;
    var recorder, gumStream;

    var handleSuccess = function (stream) {
      gumStream = stream;
      var options = {
        mimeType: 'audio/webm;codecs=opus'
      };
      recorder = new MediaRecorder(stream);
      recorder.ondataavailable = function (e) {
        var url = URL.createObjectURL(e.data);
        var preview = document.createElement('audio');
        preview.controls = true;
        preview.src = url;
        my_div.appendChild(preview);

        reader = new FileReader();
        reader.readAsDataURL(e.data);
        reader.onloadend = function () {
          base64data = reader.result;
        }
      };
      recorder.start();
    };

    recordButton.innerText = "Recording... press to stop";

    navigator.mediaDevices.getUserMedia({ audio: true }).then(handleSuccess);


    function toggleRecording() {
      if (recorder && recorder.state == "recording") {
        recordButton.innerText = "Saving the recording... please wait!"
        recordButton.style.backgroundColor = "yellow";
        recordButton.style.color = "black";
        recorder.stop();
        gumStream.getAudioTracks()[0].stop();
      }
    }

    function sleep(ms) {
      return new Promise(resolve => setTimeout(resolve, ms));
    }

    var data = new Promise(resolve => {
      recordButton.onclick = () => {
        toggleRecording()

        sleep(2000).then(() => {
          resolve(base64data.toString())
        });
      }
    });

  </script>
"""

SAVED_AUDIO_RECORDING = """
<script>
recordButton.innerText = "Audio successfully saved"
recordButton.style.backgroundColor = "green";
recordButton.style.color = "white";
</script>
"""


src/ColabKit/media.py
from .template import VIDEO_SHOW, AUDIO_SHOW
from .utils import file_exists, remove_file
from IPython.display import display, HTML
from base64 import b64encode
import ffmpeg


def showMedia(file_path, media) -> None:
    if file_exists(file_path):
        content_type = 'video/mp4' if media == "video" else 'audio/wav'
        media_type = VIDEO_SHOW if media == "video" else AUDIO_SHOW
        data = open(file_path, 'rb').read()
        data_url = f"data:{content_type};base64," + b64encode(data).decode()
        display(HTML(media_type % data_url))
    else:
        print(
            f"\033[43m\033[101mThe {media} file {file_path} is not found in the directory.\033[0m")


def get_duration(file) -> float:
    info = ffmpeg.probe(file)
    return float(info['format']['duration'])


def get_video_resolution(file) -> tuple:
    info = ffmpeg.probe(file)['streams'][0]
    if "video" == info['codec_type']:
        return (info['width'], info['height'])
    print(
        f"\033[43m\033[101mFile {file} is not a valid parameter, you must provide a video file\033[0m")


def get_formatted_error(stderr) -> str:
    lines = stderr.decode('utf-8').splitlines()
    error_lines = [line for line in lines if line.startswith("Error")]
    for error_line in error_lines:
        return f"\033[43m\033[101m{error_line}\033[0m"


def run_process(output_stream, output_file, msg) -> bool:
    try:
        ffmpeg.run(output_stream, capture_stderr=True, overwrite_output=True)
        print(f'File {msg} and saved as {output_file}')
        return True
    except ffmpeg.Error as e:
        print(get_formatted_error(e.stderr))
        remove_file(output_file)
        return False


def convert(input_file: str, output_file: str) -> bool:
    print(f"Converting {input_file} to {output_file}...")
    input_stream = ffmpeg.input(input_file)
    output_stream = ffmpeg.output(input_stream, output_file)
    return run_process(output_stream, output_file, "converted")


def shorten(input_file, start, interval, output_file) -> str:
    print(f"shorting {input_file} to {interval} seconds...")
    input_stream = ffmpeg.input(input_file, ss=start)
    output_stream = ffmpeg.output(input_stream, output_file, t=interval)
    return run_process(output_stream, output_file, "shortened")


def resize_video(video_path: str, new_resolution: int):
    new_path = f"{video_path.split('.')[0]}_{new_resolution}p.mp4"
    video_resolution = get_video_resolution(video_path)
    if video_resolution:
        print(f"Video resolution: {video_resolution}")
        if video_resolution[0] >= 1920 or video_resolution[1] >= 1080:
            print(f"Resizing video to {new_resolution}p...")
            new_width = 2 * \
                int((video_resolution[0] /
                    video_resolution[1] * new_resolution / 2))
            input_stream = ffmpeg.input(video_path)
            output_stream = ffmpeg.output(
                input_stream, new_path, vf=f'scale={new_width}:{new_resolution}', vcodec='libx264')
            return new_path if run_process(output_stream, new_path, "resized") else False
        else:
            print("No need to change the size")
            return video_path
